### PR TITLE
Update from facebook

### PR DESCRIPTION
**Pull request by pjh5**

### CODE REVIEW

File: aten/src/ATen/native/EmbeddingBag.cpp [hunk 1]
```diff
@@ -25,8 +25,9 @@ namespace native {
 
 static void make_offset2bag(const Tensor &offsets, const Tensor &indices,
                             Tensor &offset2bag) {
-  offset2bag.index_fill_(0, offsets, 1); // offset2bag = [1 0 1 0 1]
-  offset2bag[0] = 0;                     // offset2bag = [0 0 1 0 1]
+  offset2bag.index_add_(
```
Comment (soumith) [inline aten/src/ATen/native/EmbeddingBag.cpp]: was this change reviewed by someone? @colesbury did you review it?

File: torch/nn/modules/sparse.py [hunk 2]
```diff
@@ -122,7 +122,7 @@ def extra_repr(self):
         return s.format(**self.__dict__)
 
     @classmethod
-    def from_pretrained(cls, embeddings, freeze=True):
+    def from_pretrained(cls, embeddings, freeze=True, sparse=False):
```
Comment (soumith) [inline torch/nn/modules/sparse.py]: this change wasn't reviewed either I presume. Docstring wasn't updated atleast

File: test/test_nn.py [hunk 3]
```diff
@@ -1565,12 +1566,13 @@ def _test_EmbeddingBag(self, cuda, mode, sparse, dtype=torch.double):
                  [3, 4]], device=device, dtype=dtype)
 
         output = es(input, offsets)
-        output.backward(grad_output)
+        output.backward(grad_output if cuda else grad_output_with_empty)
```
Comment (soumith) [inline test/test_nn.py]: this looks wrong (maybe wrong merge conflict?). There's no `grad_output_with_empty` in this file.

**Review by soumith — CHANGES_REQUESTED** [review]

File: torch/nn/utils/clip_grad.py [hunk 4]
```diff
@@ -31,7 +31,7 @@ def clip_grad_norm_(parameters, max_norm, norm_type=2):
     clip_coef = max_norm / (total_norm + 1e-6)
     if clip_coef < 1:
         for p in parameters:
-            p.grad.data.mul_(clip_coef.item())
+            p.grad.data.mul_(float(clip_coef))
```
Comment (soumith) [inline torch/nn/utils/clip_grad.py]: this doesn't look right either, and is an unreviewed change, so has to be backed out.

File: aten/src/ATen/native/EmbeddingBag.cpp (hunk 1 above)
Comment (pjh5) [reply aten/src/ATen/native/EmbeddingBag.cpp]: I will remove this.

For context, it seems that this slipped through from D7875031, which landed before the last github->fbcode sync and therefore didn't take into account PR 5725. There is a new PR 7389 which should contain all of D7875031 and account for PR 5725, and none of the changes in D7875031 should appear in this sync. See my merge note https://github.com/pytorch/pytorch/pull/7451/commits/d6790722c98f8b5fb5fd80838b8f1cd44649a494

File: torch/nn/utils/clip_grad.py (hunk 4 above)
Comment (pjh5) [reply torch/nn/utils/clip_grad.py]: Caused by D7784132

File: torch/nn/modules/sparse.py (hunk 2 above)
Comment (pjh5) [reply torch/nn/modules/sparse.py]: Caused by D7784132

File: test/test_nn.py (hunk 3 above)
Comment (pjh5) [reply test/test_nn.py]: This also slipped through from D7875031. See the comment thread on aten/src/ATen/native/EmbeddingBag.cpp line 28

**Review by soumith — APPROVED** [review]
approved the pytorch bits

**Comment by pjh5** [comment]
Force pushed a rebase onto master and git submodule update

**Comment by pjh5** [comment]
Force pushed another rebase

**Comment by pjh5** [comment]
Force pushed an interactive rebase to get rid of revert-commits and their paired inital commits. It was possible the revert commits were causing problems by shadowing more recent changes in master, which this PR is rebased off of

**Comment by pjh5** [comment]
This PR does not currently include D7802137 or D7784132 . 
It also does not include D7875031 , but all the changes in that diff should be in https://github.com/pytorch/pytorch/pull/7389/

**Comment by pjh5** [comment]
This PR now does not include D7784132
 It also does not include D7875031 , but all the changes in that diff should be in #7389
It also does not include  D7881937 since it was reverted in D7963936

**Comment by pjh5** [comment]
Final force-push to fix authors of last 4 commits and to rebase on master

**Pull request merged**